[YUNIKORN-790] Implement MaxApplications enforcement

lixmgl opened the pull request:
### What is this PR for?
Adding maxApplication implementation for Yunikorn queue.
This PR includes and updates existing wip https://github.com/apache/yunikorn-core/pull/384 


### What type of PR is it?
* [ ] - Bug Fix
* [ ] - Improvement
* [x] - Feature
* [ ] - Documentation
* [ ] - Hot Fix
* [ ] - Refactoring

### Todos
NA

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-790

### How should this be tested?
1. local test passed by running "make test" in Yunikorn-core and Yunikorn-shim repos(refer to screenshot)

![Screen Shot 2022-09-01 at 10 30 33 PM](https://user-images.githubusercontent.com/7441350/188065791-3ad8d2d3-c9f2-4333-9b50-d9e5668cb54f.png)
![Screen Shot 2022-09-01 at 10 32 28 PM](https://user-images.githubusercontent.com/7441350/188065802-d9fe145e-8ff4-480f-8371-c801adcb1f8b.png)

2.Local test result: (launched 1000 Spark jobs to 5 Yunikorn queues on k8s cluster which has 85 nodes)
![Screen Shot 2022-08-30 at 10 27 58 PM](https://user-images.githubusercontent.com/7441350/187764274-86e2a6f9-f1ca-4da1-b4dc-7c7adb8f5778.png)

Total duration and Performance looks good.
Each Yunikorn queue set maxApplication to 12, I queried running jobs number every minute, didn't exceed the limit (12*5 queues= 60).  
[load test result@08_30.pdf](https://github.com/apache/yunikorn-core/files/9464459/load.test.result%4008_30.pdf)

### Questions:
NA

wilfred-s commented on pkg/scheduler/objects/application_state.go:
```diff
@@ -143,6 +143,7 @@ func NewAppState() *fsm.FSM {
 			},
 			fmt.Sprintf("enter_%s", Starting.String()): func(event *fsm.Event) {
 				app := event.Args[0].(*Application) //nolint:errcheck
+				app.queue.incRunningApps()
```
This is the correct point (see the [usage tracking doc](https://docs.google.com/document/d/19FYOL1Uw7tSf74qyuVN40Fv-GdWd2KrBcHJpUQLTBZ4/edit#heading=h.481iug6rohb3)) but we need to make sure the metrics are updated in the same place as the queue itself.

lixmgl replied:
Updated in https://github.com/apache/yunikorn-core/pull/435

lixmgl replied:
Updated

wilfred-s commented on pkg/scheduler/objects/queue.go:
```diff
@@ -657,6 +660,9 @@ func (sq *Queue) addChildQueue(child *Queue) error {
 	if sq.IsDraining() {
 		return fmt.Errorf("cannot add a child queue when queue is marked for deletion: %s", sq.QueuePath)
 	}
+	if sq.maxRunningApps != 0 && sq.maxRunningApps < child.maxRunningApps {
+		return fmt.Errorf("parent maxRunningApps must be larger than child maxRunningApps")
+	}
```
This belongs in the config validations check, like we do for the "other" resources. Otherwise we could fail during the config change and leave the system in a strange state.

lixmgl replied:
Can you point me where is other resources config validations check?
If you are referring this method: https://github.com/apache/yunikorn-core/blob/branch-1.0/pkg/scheduler/objects/queue.go#L207
I don't think it's a good idea to add this child maxRunningApps check into this method since we need to iterate each child.

wilfred-s commented on pkg/scheduler/objects/queue.go:
```diff
@@ -1292,3 +1304,67 @@ func (sq *Queue) String() string {
+	sq.internalIncRunningApps()
+}
+
+func (sq *Queue) internalIncRunningApps() {
+	sq.Lock()
+	defer sq.Unlock()
+	sq.runningApps++
+}
```
This can be folded into the `incRunningApps()`

lixmgl replied:
The reason I separated this logic into a new method instead of merging into incRunningApps() is: 
Based on my understanding, instruction order is not ensured in golang even with lock  (refer: https://stackoverflow.com/questions/60340382/can-i-use-lock-to-ensure-instruction-order)
If we merge into incRunningApps(), sq.Lock() could get executed before this recursion. Each parent could be locked during recursion.  
if sq.parent != nil {
		ok = sq.parent.canRun()
}

pbacsko replied:
I want to have some deeper understanding with this.

pbacsko replied:
@lixmgl I read the stackoverflow topic and that's related to goroutine scheduling. If what you're saying is correct that would mean that sequantial consistency is broken. What you're doing here is single threaded, meaning everything runs on a single goroutine. To me it would be extremely suprirising if merging `internalIncRunningApps()` into `incRunningApps()` would result in incorrect behavior.

lixmgl replied:
@pbacsko what i did is first find the parent recursively [method](https://github.com/apache/yunikorn-core/pull/429/files#diff-27632d48eb925e150a33bc92370ceaa66c31048018d11ca7a53a0b50ab7250acR1310 ) 
Then lock this parent sq only and incRunningApps. It's not single threaded. 
This way we can avoid locking all the parents during the recursion. 
I recall @wilfred-s pointed out this issue here https://github.com/apache/yunikorn-core/pull/384/files#r843616529 for combining into one method.

craigcondit replied:
Inlining `internalIncRunningApps()` will not change the behavior in any way, because the unlock will happen prior to returning from `incRunningApps()` in either case. They are functionally equivalent. However, do be sure that the lock and defer still come immediately before incrementing `runningApps`. They should not be relocated to the top of `incRunningApps()`

lixmgl replied:
@craigcondit Thanks for the info, it make sense.
I updated by merging them into one method. Please review.

wilfred-s commented on pkg/scheduler/objects/queue.go:
```diff
@@ -1292,3 +1304,67 @@ func (sq *Queue) String() string {
+	sq.internalDecRunningApps()
+}
+
+func (sq *Queue) internalDecRunningApps() {
+	sq.Lock()
+	defer sq.Unlock()
+	sq.runningApps--
+}
```
This can be folded into the `decRunningApps()`

lixmgl replied:
same reason as before.

craigcondit replied:
And as above, this is unnecessary, and these can be combined.

lixmgl replied:
Updated.

wilfred-s commented on pkg/scheduler/objects/queue.go:
```diff
@@ -1292,3 +1304,67 @@ func (sq *Queue) String() string {
+	sq.runningApps--
+}
+
+func (sq *Queue) canRun() bool {
+	if sq == nil {
+		return false
+	}
+	ok := true
+	if sq.parent != nil {
+		ok = sq.parent.canRun()
+	}
+	return sq.internalCanRun(ok)
+}
+
+func (sq *Queue) internalCanRun(ok bool) bool {
+	sq.RLock()
+	defer sq.RUnlock()
+	if sq.maxRunningApps == 0 {
+		return true && ok
+	}
+	return ok && sq.runningApps < sq.maxRunningApps
+}
```
I had to really look twice to grasp what was going on with the `ok` flag. Isn't this a much simpler flow and faster, as we fail fast if a parent has failed:
```
func (sq *Queue) canRun() bool {
	if sq == nil {
		return false
	}
	// check the parent(s)
	if sq.parent != nil {
		ok := sq.parent.canRun()
		if !ok {
			return false
		}
	}
	// check this queue only if all parents allowed it
	sq.RLock()
	defer sq.RUnlock()
	if sq.maxRunningApps == 0 {
		return true
	}
	return sq.runningApps < sq.maxRunningApps
}
```

lixmgl replied:
same reason as before, also I updated logic in canRun https://github.com/apache/yunikorn-core/pull/435

lixmgl replied:
@craigcondit Actually I removed canRun() method, this file is outdated. 
Current canRun logic is inside this method: https://github.com/apache/yunikorn-core/pull/429/commits/030c1b44a69ec8302fc566f57af7489b26d5cc0e#diff-27632d48eb925e150a33bc92370ceaa66c31048018d11ca7a53a0b50ab7250acR1345

Please review and lmk if you have any comments.

wilfred-s commented on pkg/scheduler/objects/queue.go:
```diff
@@ -1292,3 +1304,67 @@ func (sq *Queue) String() string {
+	return sq.runningApps
+}
+
+func (sq *Queue) SetMaxRunningApps(maxApps uint64) {
```
Add a comment to the exported function to show it is test only

lixmgl replied:
deleted this method since no longer needed.

wilfred-s reviewed (changes requested):
configuration check needs to move to the configvalidator code.

wilfred-s commented:
I am not sure which PR is the one you are working on. The email pointed back to this PR:
config validation code for the resources of a queue compared to its parent is here https://github.com/apache/yunikorn-core/blob/master/pkg/common/configs/configvalidator.go#L79

lixmgl commented:
> I am not sure which PR is the one you are working on. The email pointed back to this PR: config validation code for the resources of a queue compared to its parent is here https://github.com/apache/yunikorn-core/blob/master/pkg/common/configs/configvalidator.go#L79

I merged two PRs together. Please only review this one.

lixmgl commented:
> 



> 

updated, please review.

pbacsko commented on pkg/scheduler/objects/queue.go:
```diff
@@ -1292,3 +1306,83 @@ func (sq *Queue) String() string {
+	return sq.InternalIncAllocatingAcceptedAppsIfCanRun(ok, sa)
+}
+
+func (sq *Queue) InternalIncAllocatingAcceptedAppsIfCanRun(ok bool, sa *Application) bool {
```
Can we have something else instead of "ok"?

"ok" is basically an idiomatic variable for various things such as map lookup results.

lixmgl replied:
updated, using success instead of ok

pbacsko commented on pkg/scheduler/objects/queue.go:
```diff
@@ -1292,3 +1306,83 @@ func (sq *Queue) String() string {
+	}
+	result := ok && (sq.runningApps+uint64(len(sq.allocatingAcceptedApps)) < sq.maxRunningApps)
+	if result {
+		_, contains := sq.allocatingAcceptedApps[sa.ApplicationID]
```
Here, `contains` is typically named `ok`.

lixmgl replied:
updated, using ok instead of contains

pbacsko commented on pkg/scheduler/objects/application_state.go:
```diff
@@ -143,6 +143,10 @@ func NewAppState() *fsm.FSM {
 				app := event.Args[0].(*Application) //nolint:errcheck
+				app.queue.incRunningApps()
+				app.queue.decAllocatingAcceptedApps(app)
+				metrics.GetQueueMetrics(app.queuePath).IncQueueApplicationsRunning()
```
We have this call at two places now: `enter_Starting` and `enter_Running`.

lixmgl replied:
good catch, removed the one in enter_Starting, kept the one in enter_Running

pbacsko commented on pkg/scheduler/objects/application_state.go:
```diff
@@ -143,6 +143,10 @@ func NewAppState() *fsm.FSM {
+				app.queue.incRunningApps()
+				app.queue.decAllocatingAcceptedApps(app)
+				metrics.GetQueueMetrics(app.queuePath).IncQueueApplicationsRunning()
+				metrics.GetSchedulerMetrics().IncTotalApplicationsRunning()
```
Same for this, line 174 and 175.

lixmgl replied:
same

pbacsko commented on pkg/scheduler/objects/queue.go:
```diff
@@ -1292,3 +1306,83 @@ func (sq *Queue) String() string {
+	sq.Lock()
+	defer sq.Unlock()
+	if sq.maxRunningApps == 0 {
+		return true && ok
```
just `return ok`

lixmgl replied:
updated

pbacsko commented on pkg/scheduler/objects/utilities_test.go:
```diff
@@ -49,11 +49,20 @@ func createManagedQueue(parentSQ *Queue, name string, parent bool, maxRes map[st
 
 // create managed queue with props set
 func createManagedQueueWithProps(parentSQ *Queue, name string, parent bool, maxRes, props map[string]string) (*Queue, error) {
+	return createManagedQueuePropsMaxApps(parentSQ, name, parent, maxRes, props, uint64(0))
```
Does `maxApps` have to be an `uint64`? I mean I understand that it's more logical, but now we have these weird casts here and in line queue.go:1359. Simple `int` is the most portable data type.

lixmgl replied:
updated to int64

lixmgl replied:
reverted to unsigned int type. cc @wilfred-s 
related discussion is here: https://github.com/apache/yunikorn-site/pull/188#discussion_r990446069

pbacsko commented on pkg/scheduler/objects/utilities_test.go:
```diff
@@ -49,11 +49,20 @@ func createManagedQueue(parentSQ *Queue, name string, parent bool, maxRes map[st
 
 // create managed queue with props set
 func createManagedQueueWithProps(parentSQ *Queue, name string, parent bool, maxRes, props map[string]string) (*Queue, error) {
+	return createManagedQueuePropsMaxApps(parentSQ, name, parent, maxRes, props, uint64(0))
```
Does `maxApps` have to be an `uint64`? I mean I understand that it's more logical, but now we have these weird casts here and in queue.go:1359. Simple `int` is the most portable data type.

pbacsko replied:
Does `maxApps` have to be an `uint64`? I mean I understand that it's more logical, but now we have these weird casts here and in line queue.go:1359. Simple `int` is the most portable data type.

pbacsko reviewed (changes requested):
I have some concerns regarding some pieces of the code, please take a look at those, thanks.

pbacsko commented on pkg/scheduler/objects/utilities_test.go:
```diff
@@ -49,11 +49,20 @@ func createManagedQueue(parentSQ *Queue, name string, parent bool, maxRes map[st
 
 // create managed queue with props set
 func createManagedQueueWithProps(parentSQ *Queue, name string, parent bool, maxRes, props map[string]string) (*Queue, error) {
+	return createManagedQueuePropsMaxApps(parentSQ, name, parent, maxRes, props, uint64(0))
```
Does `maxApps` have to be an `uint64`? I mean I understand that it's more logical, but now we have these weird casts here and in line queue.go:1359. Simple `int` is the most portable data type.

lixmgl replied:
same

lixmgl replied:
reverted to unsigned int type. cc @wilfred-s
related discussion is here: https://github.com/apache/yunikorn-site/pull/188#discussion_r990446069

pbacsko commented on pkg/scheduler/objects/utilities_test.go:
```diff
@@ -49,11 +49,20 @@ func createManagedQueue(parentSQ *Queue, name string, parent bool, maxRes map[st
 
 // create managed queue with props set
 func createManagedQueueWithProps(parentSQ *Queue, name string, parent bool, maxRes, props map[string]string) (*Queue, error) {
+	return createManagedQueuePropsMaxApps(parentSQ, name, parent, maxRes, props, uint64(0))
```
Does `maxApps` have to be an `uint64`? I mean I understand that it's more logical, but now we have these weird casts here and in queue.go:1359. Simple `int` is the most portable data type.

lixmgl replied:
same

lixmgl replied:
same reason as before

pbacsko reviewed (changes requested):
I have some concerns regarding some pieces of the code, please take a look at those, thanks.

lixmgl commented:
> I have some concerns regarding some pieces of the code, please take a look at those, thanks.

Thanks for reviewing the code. I have updated them and addressed all your comments. Please take another look.

lixmgl commented:
Hi @manirajv06 Can you take a look? I saw @wilfred-s requested a review from you 11 days ago.

craigcondit reviewed (changes requested):
I think we're pretty close. All I see is the merge of some of the inline functions. There's no thread-safety issues with doing so, and it makes the code clearer.

Provided that is completed, I think we're ready for a merge.

lixmgl commented:
> 

Thanks @craigcondit for reviewing the PR.
I updated your comments about combining the methods. 
_canRun_ code is outdate, please review _incAllocatingAcceptedAppsIfCanRun_ method

craigcondit reviewed (changes requested):
Minor nit on log statement and one more function to combine, otherwise looks ok. @wilfred-s, @pbacsko can you take one last look at this as well?

craigcondit commented on pkg/scheduler/objects/queue.go:
```diff
@@ -1292,3 +1306,71 @@ func (sq *Queue) String() string {
+	return sq.InternalIncAllocatingAcceptedAppsIfCanRun(success, sa)
+}
+
+func (sq *Queue) InternalIncAllocatingAcceptedAppsIfCanRun(success bool, sa *Application) bool {
```
This can be combined into incAllocatingAppsIfCanRun() as well, since it is a tail call. I think that improves the visibility.

lixmgl replied:
Sure, combined.

lixmgl commented on pkg/scheduler/objects/queue.go:
```diff
@@ -1164,13 +1175,16 @@ func (sq *Queue) TryReservedAllocate(iterator func() NodeIterator) *Allocation {
+				if app.IsStarting() || app.IsRunning() || app.IsAccepted() {
+					alloc := app.tryReservedAllocate(headRoom, iterator)
+					if alloc != nil {
+						log.Logger().Debug("Rainie test accepted apps: reservation found for allocation found on queue",
```
good catch, removed :)

lixmgl commented:
> 

Thank you @craigcondit. I addressed comments.
@wilfred-s @pbacsko please take a look when you have time, thanks!

craigcondit reviewed (approved):
+1 looks good to me.

wilfred-s commented:
@lixmgl / @pbacsko / @manirajv06 / @craigcondit 
Please check PR #455. I ported the changes to master and moved the logic into the queue code.
I added unit tests and one full cycle test into the partition.
I think I have covered all the cases.

lixmgl commented:
> @lixmgl / @pbacsko / @manirajv06 / @craigcondit Please check PR #455. I ported the changes to master and moved the logic into the queue code. I added unit tests and one full cycle test into the partition. I think I have covered all the cases.

PR https://github.com/apache/yunikorn-core/pull/455 LGTM.
Thanks for the refactoring and full cycle tests @wilfred-s 
Happy late Thanksgiving :)
Can you merge my PR first then add your PR?

wilfred-s commented:
> Can we merge this PR to master first then add your PR?

This PR formed the basis under my port. We can merge the ported PR in one go. @craigcondit will handle the proper attribution to you when we commit the change. I just did some cleanup and the port :-)

lixmgl commented:
> > Can we merge this PR to master first then add your PR?
> 
> This PR formed the basis under my port. We can merge the ported PR in one go. @craigcondit will handle the proper attribution to you when we commit the change. I just did some cleanup and the port :-)

I see, sounds good. Thanks @wilfred-s and @craigcondit !

craigcondit commented:
Closed in favor of #455.
